### PR TITLE
Median Price Oracle

### PR DESCRIPTION
**Pull request by emaG3m**

### CODE REVIEW

File: contracts/Oracles/OracleManagerAdapter.sol [hunk 1]
```diff
@@ -8,25 +9,57 @@ import './TimedOracle.sol';
...
+    bytes32 result;
+    _requireValidStartIndex(startIndex);
+    _requireValidEndIndex(endIndex);
+    result = medianDataFeed.medianizeByIndices(startIndex, endIndex);
     _setOutcome(int(result));
   }
+
+  // Internal Functions
+
+  function _requireValidStartIndex(uint startIndex) internal {
```
Comment (mikec) [inline contracts/Oracles/OracleManagerAdapter.sol]: could be cleaner to use modifiers for these .. `isValidStartIndex(startIndex)`, etc

File: contracts/Oracles/OracleManagerAdapter.sol [hunk 2]
```diff
@@ -1,5 +1,6 @@
 pragma solidity ^0.4.24;
```
Comment (mikec) [inline contracts/Oracles/OracleManagerAdapter.sol]: Since this contract itself is the oracle, maybe it could just be called `MedianOracle`

Comment (emaG3m) [reply contracts/Oracles/OracleManagerAdapter.sol]: agree

File: contracts/Oracles/OracleManagerAdapter.sol [hunk 3]
```diff
@@ -8,25 +9,57 @@ import './TimedOracle.sol';
 * Aragon's Oracle Manager App which uses tidbit oracle interface
```
Comment (mikec) [inline contracts/Oracles/OracleManagerAdapter.sol]: should update the comment since this isn't directly using oracle manager app anymore

File: contracts/Oracles/OracleManagerAdapter.sol (hunk 1 above)
Comment (emaG3m) [reply contracts/Oracles/OracleManagerAdapter.sol]: this will cut down on the specificity of the revert message since now I can only put one in for 2 different cases....but I guess that's fine..they're pretty related annywayyyyyyyyy.....

Comment (emaG3m) [reply contracts/Oracles/OracleManagerAdapter.sol]: made the change even though the internal functions are not as nice

Comment (emaG3m) [reply contracts/Oracles/OracleManagerAdapter.sol]: I'll gladly change it back if you're indifferent or have anything neater :D

File: contracts/Oracles/MedianPriceOracle.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    bool endIndexWithinTimeframe = endDate <= resolutionDate;
+
+    bool lastAvailableEndIndex = true;
+    if (medianDataFeed.doesIndexExistFor(endIndex.add(1))) {
+      (, uint nextDate) = medianDataFeed.resultByIndexFor(endIndex.add(1));
+      lastAvailableEndIndex = nextDate > resolutionDate;
+    }
+
+    return endIndexWithinTimeframe && lastAvailableEndIndex;
+  }
```
Comment (mikec) [inline contracts/Oracles/MedianPriceOracle.sol]: thinking through some edge case scenarios where this check may not behave as expected, let me know what you think...

1) if no results are logged during the mediate timeframe (`resolutionDate - medianTimeframe` to `resolutionDate`), but there are results logged before and after the median timeframe, I think this check will actually pass if you send the index of the date immediately before the timeframe as the `endIndex` and the index of the date immediately after the timeframe as the `startIndex` .. so in that case it would allow an `endIndex` that is less than the `startIndex`.

2) If only 1 date is logged during the median timeframe, it is valid to pass `startIndex` and `endIndex` as the same index .. which I think is ok and produces the result we would want

3) If no dates were logged until after `resolutionDate` is passed, I think it is impossible to pass this require check, because `endDate` has to be less than or equal to resolution date. This is critical to address, since it would lock all value for scalar markets relying on this as an oracle. We might not necessarily need to change this code, it could be sufficient to just log some results after deployment of one of these median data feeds. Maybe we could add a check on initialization of this contract that there has been some data logged on `medianDataFeed` (`resultByIndexFor(1)` doesn't revert)

Comment (emaG3m) [reply contracts/Oracles/MedianPriceOracle.sol]: 1. we can do an extra check that `endIndex >= startIndex`. I didn't do it here bc it happens in the datafeed.

2. you're already able to do this if you'd like, no problems there.

3. we already have a ticket for this

Comment (mikec) [reply contracts/Oracles/MedianPriceOracle.sol]: 1) don't need to do an extra check. Since it's not allowed, this edge case is actually related to #3 I think .. if there is no data logged within the timeframe, the market will be locked.

2) an extra test for this case would be good!

3) we can merge and address in a subsequent PR then

File: contracts/Oracles/MedianPriceOracleFactory.sol [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      medianTimeframe = _medianTimeframe;
+    }
+
+    /**
+    * @dev Creates a new centralized & time-constrained oracle contract
+    * @param ipfsHash Hash idxentifying off chain event description
+    * @param resolutionDate starting date for which oracle can be resolved
+    * @return Oracle contract
+    */
+    function createOracle(bytes ipfsHash, uint resolutionDate)
```
Comment (mikec) [inline contracts/Oracles/MedianPriceOracleFactory.sol]: want to rebase and update this without the ipfsHash?

Comment (emaG3m) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: nahh I wanna rebase in the opposite order

Comment (mikec) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: kk

File: contracts/Oracles/MedianPriceOracleFactory.sol [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import './IScalarPriceOracleFactory.sol';
+
+contract MedianPriceOracleFactory is IScalarPriceOracleFactory {
+
+    event MedianPriceOracleCreation(address indexed creator, IScalarPriceOracle medianPriceOracle, bytes ipfsHash, uint resolutionDate);
+
+    address timeMedianDataFeed;
+    uint medianTimeframe;
+
+
```
Comment (mikec) [inline contracts/Oracles/MedianPriceOracleFactory.sol]: xtra space

File: contracts/Oracles/MedianPriceOracleFactory.sol (hunk 5 above)
Comment (mikec) [inline contracts/Oracles/MedianPriceOracleFactory.sol]: in a futarchy DAO, is this called by another factory contract when the futarchy markets are created, or does this need to be called externally and separate from market creation? Asking because we want to make sure that `resolutionDate` is always the same amount of time after market creation (for example, 3 weeks after creation for every market) and this should be codified at the contract level somewhere. If it needs to be called separately by the market creator, they could set this to the max uint value, and the market will never resolve .. or just set some arbitrary resolution date.

Comment (emaG3m) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: I was thinking about this a little. We'd have `futarchy-app` instantiated with a factory that gets called (just like it is right now).  We need to think about how this factory is set when the `futarchy-app` is instantiated

Comment (emaG3m) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: for now we can just give everyone a set factory with a reasonable timeframe?? until we have capcacity to customize

Comment (mikec) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: @emaG3m it looks like the `newDecision` function is already calling `createOracle` https://github.com/levelkdev/futarchy-app/blob/master/contracts/Futarchy.sol#L112 with the right resolutionDate .. always `timeToPriceResolution` after `startDate` .. thinking we just make sure that in our app init we set https://github.com/levelkdev/futarchy-app/blob/master/contracts/Futarchy.sol#L72 to an instance of `MedianPriceOracle`

Comment (mikec) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: Could also change the type here https://github.com/levelkdev/futarchy-app/blob/master/contracts/Futarchy.sol#L27 from `IScalarPriceOracleFactory` to `MedianPriceOracleFactory` .. what do you think?

Comment (emaG3m) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: yah I think that's exactly what I just said?? or are you saying something different

Comment (mikec) [reply contracts/Oracles/MedianPriceOracleFactory.sol]: cool, I guess we're thinking the same thing .. nothing in newDecision needs to change. the only things that need to change are the deployment scripts and maybe the type def?

File: contracts/Oracles/MedianPriceOracleFactory.sol [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    address timeMedianDataFeed;
+    uint medianTimeframe;
+
+    constructor(address _timeMedianDataFeed, uint _medianTimeframe) public {
+      timeMedianDataFeed = _timeMedianDataFeed;
+      medianTimeframe = _medianTimeframe;
+    }
+
+    /**
+    * @dev Creates a new centralized & time-constrained oracle contract
```
Comment (mikec) [inline contracts/Oracles/MedianPriceOracleFactory.sol]: I think this is left over from the "CentralizedTimed" oracle contract

File: test/Oracles/MedianPriceOracle.test.js [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        medianDataFeed.address,
+        MEDIAN_TIMEFRAME,
+        RESOLUTION_DATE + pingInterval,
+      )
+
+      await medianPriceOracle.setOutcome(3, 6)
+      expect((await medianPriceOracle.getOutcome()).toNumber()).to.equal(7)
+    })
+
+    it('sets the correct outcome if only one index is included in the median timeframe', async () => {
```
Comment (mikec) [inline test/Oracles/MedianPriceOracle.test.js]: coool thanks for adding this!
